nvflare/fuel/f3/cellnet/cell.py

```diff
@@ -232,26 +237,43 @@ def _get_result(self, req_id):
         waiter = self.requests_dict.pop(req_id)
         return waiter.result
 
-    def _future_wait(self, future, timeout):
+    def _check_error(self, future):
+        if future.error:
+            # must return a negative number
+            return -1
+        else:
+            return WaiterRC.OK
+
+    def _future_wait(self, future, timeout, abort_signal: Signal):
         # future could have an error!
         last_progress = 0
-        while not future.waiter.wait(timeout):
-            if future.error:
-                return False
-            current_progress = future.get_progress()
-            if last_progress == current_progress:
-                return False
+        while True:
+            rc = conditional_wait(future.waiter, timeout, abort_signal, condition_cb=self._check_error, future=future)
+            if rc == WaiterRC.IS_SET:
+                # waiter has been set!
+                break
+            elif rc == WaiterRC.TIMEOUT:
+                # timed out: check whether any progress has been made during this time
+                current_progress = future.get_progress()
+                if last_progress == current_progress:
+                    # no progress in timeout secs: consider this to be a failure
+                    return False
+                else:
+                    # good progress
+                    self.logger.debug(f"{current_progress=}")
+                    last_progress = current_progress
             else:
-                self.logger.debug(f"{current_progress=}")
-                last_progress = current_progress
+                # error condition: aborted or future error
+                return False
```

Comment:
if future has no error but timeout, the return is "WaiterRC.OK" and we go to this else statement? So OK is returning False?

Reply:
No, WaiterRC.OK is only used for condition_cb and used internally by conditional_wait, and is never returned back here!

If condition_cb (which is self._check_error) returned any thing bad (-1), it will get to this "else" block.
If the condition_cb  returns WaiterRC.OK, then we get to this "else" block only if the abort_signal is triggered.

We only check current_progress when timed out, and decide whether we should keep waiting. If there is any progress, then we'll keep waiting; otherwise returns False.